src/core/server/legacy/plugins/log_legacy_plugins_warning.ts

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Logger } from '../../logging';
+import { LegacyPluginSpec } from '../types';
+
+const internalPaths = ['/src/legacy/core_plugins', '/x-pack'];
+
+const breakingChangesUrl =
+  'https://github.com/elastic/kibana/blob/master/docs/migration/migrate_8_0.asciidoc';
+const migrationGuideUrl = 'https://github.com/elastic/kibana/blob/master/src/core/MIGRATION.md';
+
+export const logLegacyThirdPartyPluginDeprecationWarning = ({
+  specs,
+  log,
+}: {
+  specs: LegacyPluginSpec[];
+  log: Logger;
+}) => {
+  const thirdPartySpecs = specs.filter(isThirdPartyPluginSpec);
+  if (thirdPartySpecs.length > 0) {
+    const pluginIds = thirdPartySpecs.map(spec => spec.getId());
+    log.warn(
+      `Some installed third party plugin(s) [${pluginIds.join(
+        ', '
+      )}] are using the legacy plugin format and will no longer work after version 8.0. ` +
+        `Please refer to ${breakingChangesUrl} for a list of breaking changes ` +
+        `and ${migrationGuideUrl} for documentation on how to migrate legacy plugins.`
+    );
+  }
+};
+
+const isThirdPartyPluginSpec = (spec: LegacyPluginSpec): boolean => {
+  const pluginPath = spec.getPack().getPath();
+  return !internalPaths.some(internalPath => pluginPath.indexOf(internalPath) > -1);
+};
```

Comment:
So, this is not the most solid piece of software I ever crafted, but as `pack.getPath()` returns an absolute url, and as I didn't really want to work with `kbnDevUtils.REPO_ROOT` and path resolving, I felt this was sufficient, as I think it's very unlikely to have false negatives here. 

Tell me if we think this is not sufficient and if this should be improved.

Reply:
You could probably get around using kbnDevUtils by simply doing:

```ts
const rootPath = Path.resolve(Path.join(__dirname, '..', '..', '..', '..', '..'))
```

Also we may want to make the xpack path more specific `x-pack/legacy/plugins` so that any plugins loaded in functional tests `/x-pack/test/` also get the warning (similar to `test/plugin_functional`)

Reply:
> You could probably get around using kbnDevUtils by simply doing `const rootPath = Path.resolve(Path.join(__dirname, '..', '..', '..', '..', '..'))`

TBH, I'm not a big fan of relative resolves, as they are imho very fragile to refactoring. If we think this detection implementation is not good enough, I would maybe lean to still use `'@kbn/dev-utils'` as it will keep the repo root logic in a single place. OTOH the only imports of this module we currently have in `src/core/server` are in test files. So:
- Is this implementation too fragile and do we want to switch to real absolute path checking? (I'm good with that)
- If we do, should we use `@kbn/dev-utils`'s `REPO_ROOT ` or the `const rootPath = Path.resolve(Path.join(__dirname, '..', '..', '..', '..', '..'))` alternative?

> Also we may want to make the xpack path more specific `x-pack/legacy/plugins`

That would be great, except there is technically only ONE xpack `plugin`/`plugin pack`, which is `x-pack/index.js`. 

WDYT?
https://github.com/elastic/kibana/blob/982c0da78e67694e0c2e05e8e2e7431ab9bda880/x-pack/index.js#L36-L42

We don't have any information on the plugins this pack loads. I can change the xpack check to a regexp check to `/x-pack$` (or to strict absolute path checking depending on the answer of my comment's first point) however, to only target the 'real' x-pack plugin and allow to log warnings for functional test plugins.

Reply:
> * Is this implementation too fragile and do we want to switch to real absolute path checking? (I'm good with that)
> * If we do, should we use `@kbn/dev-utils`'s `REPO_ROOT ` or the `const rootPath = Path.resolve(Path.join(__dirname, '..', '..', '..', '..', '..'))` alternative?

Good point on the brittleness of refactoring. Though it's unlikely we'll refactor or move any of this legacy code in the near future, let's not create this situation if avoidable.

My main reason for not using `@kbn/dev-utils` is the fact that we don't use that module in any production code except in `src/cli`.

Given those two points, let's just keep what you have.

> We don't have any information on the plugins this pack loads. I can change the xpack check to a regexp check to `/x-pack$` (or to strict absolute path checking depending on the answer of my comment's first point) however, to only target the 'real' x-pack plugin and allow to log warnings for functional test plugins.

I forgot about the x-pack mega plugin pattern in legacy. I'm leaning towards just leaving as-is. There is only one plugin in `x-pack/test/plugin_functional` and it's a KP plugin, so this warning wouldn't apply anyways.

Reply:
you can use 
```js
const { fromRoot } = require('../../core/server/utils');
fromRoot()
```
to get absolute path to the root

> That would be great, except there is technically only ONE xpack plugin

Do we care about a warning for xpack plugins? They are internal plugins, and we can reach out to their owners directly. I think the main value of the warning is to inform 3rd party developers.